domain.c

```diff
@@ -681,103 +706,189 @@ domain_create_topleaves(int no, int * next)
-            return 1;
+    int j;
+    for(j = 0; j < 8; j ++) {
+        int sub = topTree[start].Daughter + j;
+        domain_toptree_truncate(topTree, sub, countlimit, costlimit);
+    }
+}
+
+/* remove the nodes that are no longer useful after the truncation.
+ * 
+ * We walk the topTree top-down to collect useful nodes, and move them to
+ * the head of the topTree list.
+ *
+ * This works because any child node is stored after the parent in the list 
+ * -- we are destroying the old tree just slow enough
+ * */
+
+static void
+domain_toptree_garbage_collection(struct local_topnode_data * topTree, int start, int * last_free)
+{
+
+    if(topTree[start].Daughter == -1) return;
+    int j;
+
+    int oldd = topTree[start].Daughter;
+    int newd = *last_free;
+
+    topTree[start].Daughter = newd;
+
+    (*last_free) += 8;
+    for(j = 0; j < 8; j ++) {
+        topTree[newd + j] = topTree[oldd + j];
+        topTree[newd + j].Parent = start;
+        domain_toptree_garbage_collection(topTree, newd + j, last_free);
+    }
+}
+
+/* check for local refinements with subsamples*/
+static int
+domain_check_for_local_refine_subsample(
+    struct local_topnode_data * topTree,
+    struct local_particle_data * LP,
+    int sample_step)
+{
+    NTopNodes = 1;
+    topTree[0].Daughter = -1;
+    topTree[0].Parent = -1;
+    topTree[0].Shift = BITS_PER_DIMENSION * 3;
+    topTree[0].StartKey = 0;
+    topTree[0].PIndex = 0;
+    topTree[0].Count = 0;
+    topTree[0].Cost = 0;
+
+    int i;
+
+    /* this requires the LP structure to be sorted by key;
+     * when the LP structure is sorted by key, we either refine
+     * the node the last particle lives in, or jump into a new empty node
+     * */
+
+    /* sub sample every 16 particles to create a skeleton of the toptree,
+     * because otherwise the tree is too fine!  */
+    peano_t last_key = -1;
+    int last_leaf = -1;
+
+    i = 0;
+    while(i < NumPart) {
+
+        int leaf = domain_toptree_get_subnode(topTree, LP[i].Key);
+
+        if (leaf == last_leaf) {
+            /* split last leaf to hopefully make place for the new particle
+             * we will hold at most one sample particle per leaf */
+            if(0 != domain_toptree_split(topTree, leaf)) return 1;
+            /* reinsert the last particle */
+            topTree[leaf].Count = 0;
+            last_leaf = domain_toptree_insert(topTree, last_key, 0);
+            continue;
+        } else {
+            if(topTree[leaf].Count != 0) {
+                /* this shall not happen because key is already sorted. (cite paper)*/
+                endrun(-1, "Failed to build the toptree\n");
+            }
+            last_key = LP[i].Key;
+            last_leaf = domain_toptree_insert(topTree, last_key, 0);
+            i += sample_step;
+            continue;
+        }
+    }
+    /* Now we are ready to insert all particles to the tree */
+    for(i = 0; i < NTopNodes; i ++ ) {
+        topTree[i].Count = 0;
+    }
```

Comment:
The old particles must be removed to make it a skeleton. Otherwise the counts won't be correct when we insert all particles.